plz cover doesn't link gcov when using cgo

Using `plz cover` with cgo libraries does not compile due to missing `-lgcov` linkage.

```
% plz cover //...
Build stopped after 820ms. 1 target failed:
    //foo/bar:bar_test
Error building target //foo/bar:bar_test: exit status 1
.../pkg/tool/linux_amd64/link: running /usr/bin/gcc failed: exit status 1
.../foo/bar:bar_test._build/000001.o: In function `_GLOBAL__sub_I_65535_0_link.cgo2.c':
.../foo/bar:bar_test#c_cc._build/cgo-builtin-prolog:19: undefined reference to `__gcov_init'
````

I'm able to work around this by explicitly adding linker flags.

```
cgo_library(
    name = "foo",
    srcs = [
        "bar.go",
    ],
    go_srcs = [
         "baz.go",
    ],
    linker_flags = [
        "-lgcov",
    ],
)
```

Unfortunately, this causes the cgo_library to always would link against gcov. I couldn't figure out a way to add the flag conditionally based on the presence of the cover command.

--- a/src/parse/rules/go_rules.build_defs
+++ b/src/parse/rules/go_rules.build_defs
@@ -385,7 +385,7 @@ def go_test(name:str, srcs:list, data:list=None, deps:list=[], worker:str='', vi
             'go': [CONFIG.GO_TOOL],
             'test': [CONFIG.GO_TEST_TOOL],
         },
-        post_build=lambda name, output: _replace_test_package(name, output, static, definitions=definitions),
+        post_build=lambda name, output: _replace_test_package(name, output, static, definitions=definitions, cgo=cgo),
     )
     deps.append(lib_rule)
     lib_rule = go_library(
@@ -395,7 +395,7 @@ def go_test(name:str, srcs:list, data:list=None, deps:list=[], worker:str='', vi
         _needs_transitive_deps=True,  # Rather annoyingly this is only needed for coverage
         test_only=True,
     )
-    cmds, tools = _go_binary_cmds(static=static, definitions=definitions)
+    cmds, tools = _go_binary_cmds(static=static, definitions=definitions, gcov=cgo)
 
     test_cmd = '$TEST %s 2>&1 | tee $RESULTS_FILE' % flags
     if worker:
@@ -667,7 +667,7 @@ def _go_github_repo_cmd(name, get, repo, revision):
     ], [remote_rule], [CONFIG.JARCAT_TOOL]
 
 
-def _replace_test_package(name, output, static, definitions):
+def _replace_test_package(name, output, static, definitions, cgo):
     """Post-build function, called after we template the main function.
 
     The purpose is to replace the real library with the specific one we've
@@ -683,7 +683,7 @@ def _replace_test_package(name, output, static, definitions):
             pkg_name = line[9:]
             name_changed = pkg_name != new_name
             if name_changed or ldflags or pkg_config:  # Might not be necessary if names match already.
-                binary_cmds, _ = _go_binary_cmds(static=static, ldflags=ldflags, pkg_config=pkg_config, definitions)
+                binary_cmds, _ = _go_binary_cmds(static=static, ldflags=ldflags, pkg_config=pkg_config, definitions, cgo)
                 if name_changed:
                     for k, v in binary_cmds.items():
                         set_command(new_name, k, f'mv -f $PKG_DIR/{new_name}.a $PKG_DIR/{pkg_name}.a && {v}')
@@ -714,7 +714,7 @@ def _go_library_cmds(complete=True, all_srcs=False, cover=True, filter_srcs=True
     return cmds
 
 
-def _go_binary_cmds(static=False, ldflags='', pkg_config='', definitions=None):
+def _go_binary_cmds(static=False, ldflags='', pkg_config='', definitions=None, gcov=False):
     """Returns the commands to run for linking a Go binary."""
     _link_cmd = '$TOOLS_GO tool link -tmpdir $TMP_DIR -extld $TOOLS_LD %s -o ${OUT} ' % _GOPATH.replace('-I ', '-L ')
     prefix = _LINK_PKGS_CMD + _go_import_path_cmd(CONFIG.GO_IMPORT_PATH)
@@ -741,10 +741,14 @@ def _go_binary_cmds(static=False, ldflags='', pkg_config='', definitions=None):
     if len(defs) > 0:
         flags += defs
 
-    return {
+    cmds = {
         'dbg': f'{prefix} && {_link_cmd} {flags} $SRCS',
         'opt': f'{prefix} && {_link_cmd} {flags} -s -w $SRCS',
-    }, {
+    }
+    if gcov:
+        cmds['cover'] = f'{prefix} && {_link_cmd} {flags} -extldflags="-lgcov" $SRCS'
+
+    return cmds, {
         'go': [CONFIG.GO_TOOL],
         'ld': [CONFIG.LD_TOOL if CONFIG.LINK_WITH_LD_TOOL else CONFIG.CC_TOOL],
     }

--- a/test/go_rules/cgo/BUILD
+++ b/test/go_rules/cgo/BUILD
@@ -1,9 +1,9 @@
 cgo_library(
     name = "cgo",
     srcs = ["cgo.go"],
-    go_srcs = ["go.go"],
-    c_srcs = ["cgo.c"],
     hdrs = ["cgo.h"],
+    c_srcs = ["cgo.c"],
+    go_srcs = ["go.go"],
 )
 
 cgo_test(

--- a/test/go_rules/cgo/BUILD
+++ b/test/go_rules/cgo/BUILD
@@ -1,0 +1,7 @@
+cgo_library(
+    name = "cgo",
+    srcs = ["cgo.go"],
+    go_srcs = ["go.go"],
+    c_srcs = ["cgo.c"],
+    hdrs = ["cgo.h"],
+)

--- a/test/go_rules/cgo/BUILD
+++ b/test/go_rules/cgo/BUILD
@@ -5,3 +5,12 @@ cgo_library(
     c_srcs = ["cgo.c"],
     hdrs = ["cgo.h"],
 )
+
+cgo_test(
+    name = "cgo_test",
+    srcs = ["cgo_test.go"],
+    deps = [
+        ":cgo",
+        "//third_party/go:testify",
+    ],
+)

--- a/test/go_rules/cgo/cgo.c
+++ b/test/go_rules/cgo/cgo.c
@@ -1,0 +1,5 @@
+#include "cgo.h"
+
+int GetAnswer() {
+  return 42;
+}

--- a/test/go_rules/cgo/cgo.go
+++ b/test/go_rules/cgo/cgo.go
@@ -1,0 +1,9 @@
+package cgo
+
+// #include "cgo.h"
+import "C"
+
+// GetAnswer returns the answer to the great question of Life, the Universe and Everything.
+func GetAnswer() int {
+	return int(C.GetAnswer())
+}

--- a/test/go_rules/cgo/cgo.h
+++ b/test/go_rules/cgo/cgo.h
@@ -1,0 +1,2 @@
+// GetAnswer returns the answer to the great question of Life, the Universe and Everything.
+int GetAnswer();

--- a/test/go_rules/cgo/cgo_test.go
+++ b/test/go_rules/cgo/cgo_test.go
@@ -1,0 +1,16 @@
+package cgo
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestAnswer(t *testing.T) {
+	assert.Equal(t, 42, GetAnswer())
+}
+
+func TestQuestion(t *testing.T) {
+	// There has always been something wrong with the world.
+	assert.Equal(t, "What do you get if you multiply six by nine?", GetQuestion())
+}

--- a/test/go_rules/cgo/cgo_test.go
+++ b/test/go_rules/cgo/cgo_test.go
@@ -11,6 +11,5 @@ func TestAnswer(t *testing.T) {
 }
 
 func TestQuestion(t *testing.T) {
-	// There has always been something wrong with the world.
-	assert.Equal(t, "What do you get if you multiply six by nine?", GetQuestion())
+	assert.NoError(t, CheckAnswer("42"))
 }

--- a/test/go_rules/cgo/go.go
+++ b/test/go_rules/cgo/go.go
@@ -1,0 +1,6 @@
+package cgo
+
+// GetQuestion returns the great Question.
+func GetQuestion() string {
+	return "What do you get if you multiply six by nine?"
+}

--- a/test/go_rules/cgo/go.go
+++ b/test/go_rules/cgo/go.go
@@ -1,6 +1,15 @@
 package cgo
 
-// GetQuestion returns the great Question.
-func GetQuestion() string {
-	return "What do you get if you multiply six by nine?"
+import (
+	"fmt"
+	"strconv"
+)
+
+// CheckAnswer checks that the given answer matches the canonical one.
+func CheckAnswer(answer string) error {
+	ans, _ := strconv.ParseInt(answer, 13, 32)
+	if int(ans) != 6*9 {
+		return fmt.Errorf("universe parameters incorrect")
+	}
+	return nil
 }

--- a/tools/jarcat/ar/ar.go
+++ b/tools/jarcat/ar/ar.go
@@ -80,7 +80,9 @@ func Create(srcs []string, out string, combine, rename bool) error {
 				hdr.ModTime = mtime
 				hdr.Uid = 0
 				hdr.Gid = 0
-				log.Debug("copying '%s' in from %s", hdr.Name, src)
+				// Fix weird bug about octal numbers (looks like we're prepending 100 multiple times)
+				hdr.Mode &= ^0100000
+				log.Debug("copying '%s' in from %s, mode %x", hdr.Name, src, hdr.Mode)
 				if err := w.WriteHeader(hdr); err != nil {
 					return err
 				} else if _, err = io.Copy(w, r); err != nil {

